File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque
N = int(sys.stdin.readline().strip())


nx = [1, 2, 2, 1, -1, -2, -2, -1]
ny = [-2, -1, 1, 2, -2, -1, 1, 2]

def BFS(x,y):

    q= deque()

    q.append([x,y])

    while q:
        fromnode = q.popleft()
        fromx = fromnode[0]
        fromy = fromnode[1]

        if (fromx == endx and fromy == endy) :
            return graph[endx][endy]

        
        for i in range(8):
            dx = fromx + nx[i]
            dy = fromy + ny[i]
            
            if 0 <= dx < length and 0 <= dy <length :
                if graph[dx][dy] == 0:
                    q.append([dx, dy])
                    graph[dx][dy] = graph[fromx][fromy] + 1  # 수정된 부분
                    
                        


for i in range(N):
    length = int(sys.stdin.readline())
    startx, starty = map(int, sys.stdin.readline().split())
    endx, endy = map(int, sys.stdin.readline().split())

    graph = [[0] * (length + 1) for _ in range(length + 1)]
    print(BFS(startx, starty))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter (new OutputStreamWriter(System.out));

        int T = Integer.parseInt(br.readLine());

        int [] dx = {-2,-1,1,2,2,1,-1,-2};
        int [] dy = {1,2,2,1,-1,-2,-2,-1};

        for (int z = 0 ; z<T ; z++)
            {
                int L = Integer.parseInt(br.readLine());
                String [] oneline = br.readLine().split(" ");
                int startx = Integer.parseInt(oneline[0]);
                int starty = Integer.parseInt(oneline[1]);

                String [] twoline = br.readLine().split(" ");
                int endx = Integer.parseInt(twoline[0]);
                int endy = Integer.parseInt(twoline[1]);

                Queue<int []> q = new LinkedList<>();
                int [] firstarr = {startx,starty, 0};
                q.add(firstarr);

                boolean[][] visited = new boolean[L][L];
                visited[startx][starty] = true;

                while(!q.isEmpty())
                    {
                        int [] firstq = q.poll();
                        int x = firstq[0];
                        int y = firstq[1];
                        int depth = firstq[2];

                        if(x == endx && y == endy)
                        {
                            bw.write(String.valueOf(depth)+"\n");
                            break;
                        }

                        for (int i = 0 ; i< 8; i++)
                            {
                                int nx = x+dx[i];
                                int ny = y+dy[i];

                                if(0<=nx && nx < L && 0<=ny && ny< L)
                                {
                                    if(!visited[nx][ny])
                                    {
                                        visited[nx][ny] = true;
                                        int [] newq = {nx,ny,depth+1};
                                        q.add(newq);
                                    }
                                }
                            }
                    }

            }
        bw.close();
    }
}